### PR TITLE
feat: Add prefer-object-has-own rule.

### LINKED ISSUE #14939: A rule suggesting to replace `Object.prototype.hasOwnProperty.call` with `Object.hasOwn`
**Please describe what the rule should do:**
Starting V8 v.9.3, `Object.prototype.hasOwnProperty.call` can be replaced with an alias/syntax sugar `Object.hasOwn`, which is much more read-friendly. Further information: https://v8.dev/features/object-has-own

**What new ECMAScript feature does this rule relate to?**
Promoting using of an alias/syntax sugar `Object.hasOwn` instead of `Object.prototype.hasOwnProperty.call`.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[X] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**
The original code:
```js
if (Object.prototype.hasOwnProperty.call(object, 'foo')) {
  // `object` has property `foo`.
}
```

Should be replaced with:
```js
if (Object.hasOwn(object, 'foo')) {
  // `object` has property `foo`.
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**
Applying this rule you can make your application's code cleaner.

**Are you willing to submit a pull request to implement this rule?**
No.

### PR DESCRIPTION
**Pull request by Gautam-Arora24**
Fixes #14939

#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[X] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)
This PR focuses on adding a new rule called `prefer-object-has-own`. The PR adheres to the guidelines provided and has tests and documentation included. This rule requires using `Object.hasOwn` instead of `Object.prototype.hasOwnProperty`.

#### Is there anything you'd like reviewers to focus on?
**Caveats ->**
The PR doesn't contain any fixes for the rule.

Fixes #14939

### CODE REVIEW

File: docs/rules/prefer-object-has-own.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# Prefer use of Object.hasOwn over `Object.prototype.hasOwnPrototype` (prefer-object-has-own)
+
+When Object.prototype.hasOwnPrototype.call is used, this rule requires using the `Object.hasOwn` instead. `Object.hasOwn` is a syntactic sugar and makes the code cleaner.
```
Comment (nzakas) [inline docs/rules/prefer-object-has-own.md:3]: This section needs to explain not just what the rule does but why it does it.

File: lib/rules/prefer-object-has-own.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
+/**
+ * @fileoverview Prefers Object.hasOwn instead of Object.prototype.hasOwnProperty
+ * @author Gautam Arora <gautamarora6248@gmail.com>
+ * See LICENSE file in root directory for full license.
```
Comment (nzakas) [inline lib/rules/prefer-object-has-own.js:4]: This line is unnecessary.

File: tests/lib/rules/prefer-object-has-own.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
+/**
+ * @fileoverview Prefers Object.hasOwn instead of Object.prototype.hasOwnProperty
+ * @author Gautam Arora
+ * See LICENSE file in root directory for full license.
```
Comment (nzakas) [inline tests/lib/rules/prefer-object-has-own.js:4]: This line is unnecessary.

File: tests/lib/rules/prefer-object-has-own.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ruleTester.run("prefer-object-has-own", rule, {
+    valid: [
+        `
+        let obj = {};
+        Object.hasOwn(obj,"");
+        `
+    ],
+    invalid: [
+        {
+            code: "Object.prototype.hasOwnProperty",
```
Comment (nzakas) [inline tests/lib/rules/prefer-object-has-own.js:36]: I think this shouldn’t cause an error.

Comment (Gautam-Arora24) [reply tests/lib/rules/prefer-object-has-own.js:36]: We are warning the developers not to use `hasOwnProperty` right? So shouldn't it cause an error if someone uses  it?

Comment (mdjermanovic) [reply tests/lib/rules/prefer-object-has-own.js:36]: > We are warning the developers not to use `hasOwnProperty` right?

We are warning on `Object.prototype.hasOwnProperty.call(obj, prop)`.

Consider this example:

```js
import { doSomething } from "foo";

doSomething(Object.prototype.hasOwnProperty);
```

If the rule warns, what can the user do to fix this code?

Comment (Gautam-Arora24) [reply tests/lib/rules/prefer-object-has-own.js:36]: Okay! Got it. We should warn only on those cases when `Object.prototype.hasOwnProperty` is actually used with call method.

File: tests/lib/rules/prefer-object-has-own.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+const parserOptions = {
+    ecmaVersion: 2022
+};
+
+const ruleTester = new RuleTester({ parserOptions });
+const error = { messageId: "useHasOwn" };
+
+ruleTester.run("prefer-object-has-own", rule, {
+    valid: [
```
[the remark this answers is not in the document]
Comment (Gautam-Arora24) [reply tests/lib/rules/prefer-object-has-own.js:28]: Sure! Will be adding these test cases as well.

**Comment by nzakas** [comment]
@Gautam-Arora24 are you still working on this?

**Comment by Gautam-Arora24** [comment]
Sorry @nzakas for the delay. I will create followup commits soon :)

**Comment by Gautam-Arora24** [comment]
Sorry @nzakas but I am not getting enough time for this PR. If someone wants to pick it up, he/she can feel free to do so :)

**Comment by snitin315** [comment]
I will fininsh it 👍

**Comment by snitin315** [comment]
closing in favor of the new PR https://github.com/eslint/eslint/pull/15206

**Comment by ljharb** [comment]
I think you meant #15346

**Comment by snitin315** [comment]
> I think you meant #15346

Definitely 👍

**Pull request closed without merging**